src/wp-includes/html-api/class-wp-html-open-elements.php

```diff
@@ -101,6 +101,49 @@ public function set_push_handler( Closure $handler ) {
 		$this->push_handler = $handler;
 	}
 
+	/**
+	 * Returns the name of the node at the nth position on the stack
+	 * of open elements, or `null` if no such position exists.
+	 *
+	 * Note that this uses a 1-based index, which represents the
+	 * "nth item" on the stack, counting from the top, where the
+	 * top-most element is the 1st, the second is the 2nd, etc...
+	 *
+	 * @todo Skip over "marker" entries on the stack, if appropriate.
+	 *
+	 * @param int $nth Retrieve the nth item on the stack, with 1 being
+	 *                 the top element, 2 being the second, etc...
+	 * @return string|null Name of the node on the stack at the given location,
+	 *                     or `null` if the location isn't on the stack.
+	 */
+	public function at( int $nth ) {
+		foreach ( $this->walk_down() as $item ) {
+			if ( 0 === --$nth ) {
+				return $item->node_name;
+			}
+		}
+
+		return null;
+	}
```

Comment:
I do want to be careful about conflating the language from other programming languages and their array data structures and the stack in the HTML spec. maybe `at()` is the wrong name. I originally added this before when I needed to implement another behavior.

perhaps we call it `stack_position( $nth )`

I like your idea for negative numbers, though maybe we could make it even more explicit.

```php
'OPTGROUP' === $this->state->stack_of_open_elements->at_position( 'from-bottom', 2 );
```

Reply:
It's true this could be confusing to use a single method and negative or positive indexes to move from either end, it may not be clear which end is which.

Instead of a single method with some better-named flag, let's match the `walk_up` (start at bottom and go towards top) and `walk_down` (start at top and go towards bottom) which already exist and match the specification stack:

```php
'DIV' === $this->state->stack_of_open_elements->nth_from_bottom( 1 ); // analogous to current node on the stack
'HTML' === $this->state->stack_of_open_elements->nth_from_top( 1 ); // always true
```

Reply:
this is great, but I think it would be best on its own, so maybe we can follow-up with a change that modifies things in a unified commit. with the other major open PRs, perhaps we can address this once we are basically at full support, or at least until the open PRs are closed.

Reply:
scratch that. I was playing with it and I'm going to incorporate it here. let's see how we go. there's a part of me that doesn't like adding more index-based methods (as I write it out), but given the spec uses this language, that's fine I guess.

Reply:
It still feels odd, so I'm going to leave this out for now and ask you your thoughts. For most cases of `walk_up()` and `walk_down()` we're referencing another node and not an index, which seems far less likely to be broken by updates to the code. For example, setting `$parent = $->nth_from_bottom(2)` depends on the current node being at position `nth = 1`. With `->walk_up( $node ) as $parent` we don't have that same fragility by exporting the internals of the stack to the HTML Processor.

I'm open to this, but the gains didn't feel as big as I expected when I went through what you proposed and did my own search to find all the cases.

```diff
diff --git a/src/wp-includes/html-api/class-wp-html-open-elements.php b/src/wp-includes/html-api/class-wp-html-open-elements.php
index 0cd223657a..6f76265b09 100644
--- a/src/wp-includes/html-api/class-wp-html-open-elements.php
+++ b/src/wp-includes/html-api/class-wp-html-open-elements.php
@@ -109,8 +109,6 @@ class WP_HTML_Open_Elements {
 	 * "nth item" on the stack, counting from the top, where the
 	 * top-most element is the 1st, the second is the 2nd, etc...
 	 *
-	 * @todo Skip over "marker" entries on the stack, if appropriate.
-	 *
 	 * @since 6.7.0
 	 *
 	 * @param int $nth Retrieve the nth item on the stack, with 1 being
@@ -128,6 +126,53 @@ class WP_HTML_Open_Elements {
 		return null;
 	}
 
+	/**
+	 * Returns the name of the node at the nth position from the bottom on
+	 * the stack of open elements, or `null` if no such position exists.
+	 *
+	 * Note that this uses a 1-based index, which represents the
+	 * "nth item" on the stack, counting from the bottom, where the
+	 * bottom-most (or deepest) element is the 1st, the second is
+	 * the 2nd, etc...
+	 *
+	 * @since 6.7.0
+	 *
+	 * @param int $nth Retrieve the nth item on the stack, with 1 being
+	 *                 the bottom (depeest) element, 2 being the second, etc...
+	 * @return string|null Name of the node on the stack at the given location,
+	 *                     or `null` if the location isn't on the stack.
+	 */
+	public function nth_from_bottom( int $nth ): ?string {
+		if ( $nth <= 0 || $nth > count( $this->stack ) ) {
+			return null;
+		}
+
+		return $this->stack[ count( $this->stack ) - $nth ]->node_name;
+	}
+
+	/**
+	 * Returns the name of the node at the nth position from the top on
+	 * the stack of open elements, or `null` if no such position exists.
+	 *
+	 * Note that this uses a 1-based index, which represents the
+	 * "nth item" on the stack, counting from the top, where the
+	 * top-most element is the 1st, the second is the 2nd, etc...
+	 *
+	 * @since 6.7.0
+	 *
+	 * @param int $nth Retrieve the nth item on the stack, with 1 being
+	 *                 the top element, 2 being the second, etc...
+	 * @return string|null Name of the node on the stack at the given location,
+	 *                     or `null` if the location isn't on the stack.
+	 */
+	public function nth_from_top( int $nth ): ?string {
+		if ( $nth <= 0 || $nth > count( $this->stack ) ) {
+			return null;
+		}
+
+		return $this->stack[ $nth - 1 ]->node_name;
+	}
+
 	/**
 	 * Reports if a node of a given name is in the stack of open elements.
 	 *
diff --git a/src/wp-includes/html-api/class-wp-html-processor.php b/src/wp-includes/html-api/class-wp-html-processor.php
index f66498b19f..0f2135387e 100644
--- a/src/wp-includes/html-api/class-wp-html-processor.php
+++ b/src/wp-includes/html-api/class-wp-html-processor.php
@@ -1211,7 +1211,7 @@ class WP_HTML_Processor extends WP_HTML_Tag_Processor {
 			case '+BODY':
 				if (
 					1 === $this->state->stack_of_open_elements->count() ||
-					'BODY' !== $this->state->stack_of_open_elements->at( 2 ) ||
+					'BODY' !== $this->state->stack_of_open_elements->nth_from_top( 2 ) ||
 					$this->state->stack_of_open_elements->contains( 'TEMPLATE' )
 				) {
 					// Ignore the token.
@@ -1237,7 +1237,7 @@ class WP_HTML_Processor extends WP_HTML_Tag_Processor {
 			case '+FRAMESET':
 				if (
 					1 === $this->state->stack_of_open_elements->count() ||
-					'BODY' !== $this->state->stack_of_open_elements->at( 2 ) ||
+					'BODY' !== $this->state->stack_of_open_elements->nth_from_top( 2 ) ||
 					false === $this->state->frameset_ok
 				) {
 					// Ignore the token.
@@ -2358,10 +2358,8 @@ class WP_HTML_Processor extends WP_HTML_Tag_Processor {
 			case '-OPTGROUP':
 				$current_node = $this->state->stack_of_open_elements->current_node();
 				if ( $current_node && 'OPTION' === $current_node->node_name ) {
-					foreach ( $this->state->stack_of_open_elements->walk_up( $current_node ) as $parent ) {
-						break;
-					}
-					if ( $parent && 'OPTGROUP' === $parent->node_name ) {
+					$parent = $this->state->stack_of_open_elements->nth_from_bottom( 2 );
+					if ( 'OPTGROUP' === $parent ) {
 						$this->state->stack_of_open_elements->pop();
 					}
 				}
@@ -3338,10 +3336,7 @@ class WP_HTML_Processor extends WP_HTML_Tag_Processor {
 	 */
 	public function reset_insertion_mode(): void {
 		// Set the first node.
-		$first_node = null;
-		foreach ( $this->state->stack_of_open_elements->walk_down() as $first_node ) {
-			break;
-		}
+		$first_node = $this->state->stack_of_open_elements->nth_from_top( 1 );
 
 		/*
 		 * > 1. Let _last_ be false.
```

Reply:
Strangely, with that patch applied I get one more skipped test/one fewer assertions. I'm not sure why, but I'm thinking it must be a logic error, which only adds to my skepticism of replacing semantic relations with array indices.

Reply:
Let's leave it out for now and revisit after this lands, that's fine.